File: android/composeui/src/main/java/com/stadiamaps/ferrostar/composeui/theme/ArrivalViewTheme.kt
package com.stadiamaps.ferrostar.composeui.theme

import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight

enum class ArrivalViewStyle {
  /** A simple arrival view with only values. */
  SIMPLIFIED,
  /** An arrival view with label captions in addition to values. */
  INFORMATIONAL
}

/** Themes for arrival view components */
interface ArrivalViewTheme {
  /** The text style for the step distance (or distance to step). */
  @get:Composable val style: ArrivalViewStyle
  /** The color for the measurement/value. */
  @get:Composable val measurementColor: Color
  /** The text style for the measurement/value. */
  @get:Composable val measurementTextStyle: TextStyle
  /** The color for the secondary content (label caption). */
  @get:Composable val secondaryColor: Color
  /** The text style for the secondary content (label caption). */
  @get:Composable val secondaryTextStyle: TextStyle
  /** The exit button icon color. */
  @get:Composable val exitIconColor: Color
  /** The exit button background color. */
  @get:Composable val exitButtonBackgroundColor: Color
  /** The background color for the view. */
  @get:Composable val backgroundColor: Color
}

object DefaultArrivalViewTheme : ArrivalViewTheme {
  override val style: ArrivalViewStyle
    @Composable get() = ArrivalViewStyle.SIMPLIFIED

  override val measurementColor: Color
    @Composable get() = MaterialTheme.colorScheme.onBackground

  override val measurementTextStyle: TextStyle
    @Composable get() = MaterialTheme.typography.titleLarge.copy(fontWeight = FontWeight.SemiBold)

  override val secondaryColor: Color
    @Composable get() = MaterialTheme.colorScheme.secondary

  override val secondaryTextStyle: TextStyle
    @Composable get() = MaterialTheme.typography.labelSmall

  override val exitIconColor: Color
    @Composable get() = MaterialTheme.colorScheme.onSecondary

  override val exitButtonBackgroundColor: Color
    @Composable get() = MaterialTheme.colorScheme.secondary

  override val backgroundColor: Color
    @Composable get() = MaterialTheme.colorScheme.background
}

File: apple/Sources/FerrostarSwiftUI/Theme/ArrivalViewTheme.swift
import SwiftUI

public enum ArrivalViewStyle: Equatable {
    /// The simplified/default which only shows actual values
    case simplified

    /// An expanded informational arrival view that labels each value.
    case informational
}

public protocol ArrivalViewTheme: Equatable {
    /// The style of the arrival view controls the general theme.
    var style: ArrivalViewStyle { get }

    /// The color for the measurement values (top row)
    var measurementColor: Color { get }

    /// The font for the measurement values (top row)
    var measurementFont: Font { get }

    /// The color for the secondary text.
    var secondaryColor: Color { get }

    /// The font for the secondary text.
    var secondaryFont: Font { get }

    /// The color of the background.
    var backgroundColor: Color { get }
}

public struct DefaultArrivalViewTheme: ArrivalViewTheme {
    public var style: ArrivalViewStyle = .simplified
    public var measurementColor: Color = .primary
    public var measurementFont: Font = .title2.bold()
    public var secondaryColor: Color = .secondary
    public var secondaryFont: Font = .subheadline
    public var backgroundColor: Color = .init(.systemBackground)

    public init() {}
}
